uniswap-v2-twap/src/UniswapV2Twap.sol

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: MIT
+// WARNING! This smart contract and the associated zk-SNARK verifiers have not been audited.
+// DO NOT USE THIS CONTRACT FOR PRODUCTION
+pragma solidity 0.8.19;
+
+import {Ownable} from "openzeppelin-contracts/access/Ownable.sol";
+import {IAxiomV1Query} from "axiom-contracts/contracts/interfaces/IAxiomV1Query.sol";
+import "utils/RLPReader.sol";
+
+contract UniswapV2Twap is Ownable {
+    using RLPReader for RLPReader.RLPItem;
+    using RLPReader for RLPReader.Iterator;
+    using RLPReader for bytes;
+
+    address public axiomQueryAddress;
+    mapping(bytes28 => uint256) public twapPris;
+
+    event UniswapV2TwapProof(
+        address pairAddress,
+        uint32 startBlockNumber,
+        uint32 endBlockNumber,
+        uint256 twapPri
+    );
+    event UpdateAxiomAddress(address newAddress);
+
+    constructor(address _axiomQueryAddress) {
+        axiomQueryAddress = _axiomQueryAddress;
+    }
+
+    function updateAxiomQueryAddress(
+        address _axiomQueryAddress
+    ) external onlyOwner {
+        axiomQueryAddress = _axiomQueryAddress;
+        emit UpdateAxiomAddress(_axiomQueryAddress);
+    }
+
+    function _getTimestampFromBlock(
+        uint32 blockNumber,
+        bytes32 blockHash,
+        bytes memory rlpEncodedHeader
+    ) internal pure returns (uint256) {
+        require(keccak256(rlpEncodedHeader) == blockHash, "invalid blockhash");
+
+        RLPReader.RLPItem[] memory ls = rlpEncodedHeader.toRlpItem().toList();
+        require(blockNumber == ls[8].toUint(), "invalid block number");
+        uint256 timestamp = ls[11].toUint();
+        return timestamp;
+    }
+
+    function _currentCumulativePrice(
+        uint112 reserve0,
+        uint112 reserve1,
+        uint256 blockTimestamp,
+        uint32 blockTimestampLast,
+        uint priceCumulativeLast
+    ) internal pure returns (uint256) {
+        //overflow is desired
+        unchecked {
+            uint256 increment = uint(reserve1 / reserve0) *
```

Comment:
Need to use `FixedPoint` fractions, see: https://github.com/Uniswap/v2-periphery/blob/0335e8f7e1bd1e8d8329fd300aea2ef2f36dd19f/contracts/libraries/UniswapV2OracleLibrary.sol#L30

Reply:
done using `UQ112x112.sol` as done here https://github.com/LongFinance/longswap-contracts/blob/a1ecf7fc341bd4c1c0dbd40cfdd2a04985d78d19/contracts/uniswapv2/UniswapV2Pair.sol#L84